src/producer/base_producer.rs

```diff
@@ -494,26 +503,26 @@ where
     // As this library uses the rdkafka Event API, flush will not call rd_kafka_poll() but instead wait for
     // the librdkafka-handled message count to reach zero. Runs until value reaches zero or timeout.
     fn flush<T: Into<Timeout>>(&self, timeout: T) -> KafkaResult<()> {
-        let mut timeout = timeout.into();
-        loop {
-            let op_timeout = std::cmp::min(timeout, self.min_poll_interval);
-            if self.in_flight_count() > 0 {
-                unsafe { rdsys::rd_kafka_flush(self.native_ptr(), 0) };
-                self.poll(op_timeout);
+        let deadline = if let Timeout::After(dur) = timeout.into() {
+            Deadline::new(dur)
+        } else {
+            Deadline::new(Duration::MAX)
+        };
+        while self.in_flight_count() > 0 && !deadline.elapsed() {
+            let ret = unsafe {
+                // This cast to i32 will truncate to i32::MAX
```

Comment:
I'm not sure about it. `Duration::as_millis()` return `u128` (https://doc.rust-lang.org/std/time/struct.Duration.html#method.as_millis). So I think if you cast something like i32::MAX + 1 then you'll get negative.

```rust
use std::time::Duration;

fn main() {
    let d = Duration::from_millis(2_147_483_647 /* i32::Max */ + 1);
    println!("{}", d.as_millis() as i32);
}
```

output: -2147483648


https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f2cda565c0c554ef158e5ea3c9a17921

Reply:
Wow; that's really surprising. I will update and handle explicitly.